payments-core/src/main/java/com/stripe/android/paymentsheet/PaymentSheetViewModel.kt

```diff
@@ -192,6 +184,47 @@ internal class PaymentSheetViewModel @Inject internal constructor(
                 stripeIntentValidator.requireValid(stripeIntent)
             }.fold(
                 onSuccess = {
+                    updatePaymentMethods(stripeIntent)
+                    resetViewState(stripeIntent, userErrorMessage = null)
+                },
+                onFailure = ::onFatal
+            )
+        }
+    }
+
+    /**
+     * Fetch the saved payment methods for the customer, if a [PaymentSheet.CustomerConfiguration]
+     * was provided.
+     * It will fetch only the payment method types accepted by the [stripeIntent] and defined in
+     * [SupportedSavedPaymentMethod].
+     */
+    @VisibleForTesting
+    fun updatePaymentMethods(stripeIntent: StripeIntent) {
+        viewModelScope.launch {
+            runCatching {
+                customerConfig?.let { customerConfig ->
+                    stripeIntent.paymentMethodTypes.mapNotNull {
+                        SupportedSavedPaymentMethod.fromCode(it)
+                    }.map {
+                        async {
+                            paymentMethodsRepository.get(
+                                customerConfig,
+                                it.type
+                            )
+                        }
+                    }.awaitAll().flatten().filter { paymentMethod ->
+                        SupportedSavedPaymentMethod.isValid(paymentMethod).also { valid ->
```

Comment:
You can remove that check on line 215 of PaymentSheetViewModel.  PaymentOptionAdapter and PaymentOptionFactory both call getSavedPaymentMethodIcon and getLabel, both of which check for null on the card, and this removes the need for extra per-type checking.